clang/lib/CodeGen/CodeGenFunction.cpp

```diff
@@ -1093,6 +1093,36 @@ void CodeGenFunction::StartFunction(GlobalDecl GD, QualType RetTy,
                     llvm::MDNode::get(getLLVMContext(), AttrMDArgs));
   }
 
+  if (getLangOpts().SYCLIsDevice && D &&
+      D->hasAttr<SYCLAddIRAttributesFunctionAttr>()) {
+    const auto *A = D->getAttr<SYCLAddIRAttributesFunctionAttr>();
+    const auto NameValuePairs = CGM.getFilteredValidAttributeNameValuePairs(
+        A->args_begin(), A->args_size(), A);
+
+    llvm::AttrBuilder FnAttrBuilder(Fn->getContext());
+    for (const auto &NameValuePair : NameValuePairs)
+      FnAttrBuilder.addAttribute(NameValuePair.first, NameValuePair.second);
+    Fn->addFnAttrs(FnAttrBuilder);
+  }
+
+  if (getLangOpts().SYCLIsDevice && D) {
+    for (unsigned I = 0; I < Args.size(); ++I) {
```

Comment:
I think this code can be moved to `CodeGenFunction::EmitFunctionProlog`. We iterate through function arguments there already. `SYCLAccessorReadonlyAttr` is handled there.

Reply:
Thank you for pointing me to this. I agree, it makes perfect sense to have it there instead. It has been moved.